python/lsst/analysis/tools/tasks/sourceObjectTableAnalysis.py

```diff
@@ -98,35 +241,147 @@ class SourceObjectTableAnalysisConfig(
         dtype=str,
         default=["pixelFlags_saturated", "pixelFlags_saturatedCenter"],
     )
+    applyProperMotions = pexConfig.Field(
+        dtype=bool,
+        default=True,
+        doc="Apply proper motions to source positions.",
+    )
 
 
 class SourceObjectTableAnalysisTask(AnalysisPipelineTask):
     ConfigClass = SourceObjectTableAnalysisConfig
     _DefaultName = "sourceObjectTableAnalysis"
 
-    def runQuantum(self, butlerQC, inputRefs, outputRefs):
-        inputs = butlerQC.get(inputRefs)
+    def callback(self, inputs, dataId):
+        """Callback function to be used with reconstructor."""
+        return self.prepareAssociatedSources(
+            dataId["visit"],
+            inputs["data"],
+            inputs["associatedSources"],
+            inputs["refCat"],
+            inputs["visitTable"],
+            inputs["properMotionCatalog"],
+        )
 
-        # Get isolated sources:
-        visit = inputs["data"].dataId["visit"]
-        band = inputs["data"].dataId["band"]
-        names = self.collectInputNames()
-        names -= {self.config.ra_column, self.config.dec_column}
-        data = inputs["data"].get(parameters={"columns": names})
+    def applyProperMotionCorrections(
+        self, isolatedSources, properMotionCatalog, visitTable, visit, refEpochs
+    ):
+        """Shift source positions to match the epoch of the reference catalog
+        objects.
 
-        dataId = butlerQC.quantum.dataId
-        plotInfo = self.parsePlotInfo(inputs, dataId)
+        Parameters
+        ----------
+        isolatedSources : `astropy.table.Table`
+            Catalog of sources to which proper motion corrections will be made.
+        properMotionCatalog : `pd.DataFrame`
+            Catalog with proper motion and parallax information.
+        visitTable : `pd.DataFrame`
+            Catalog containing the epoch for the visit corresponding to the
+            isolatedSources.
+        visit : `int`
+            Identifier of the isolatedSources' visit.
+        """
+        sourceMjd = visitTable.loc[visit]["expMidptMJD"]
+
+        # Get target date from reference catalog
+        targetEpochs = refEpochs.to_numpy()
+        # There may not be a valid reference epoch on the edge of a given
+        # region. Do not make a proper motion correction for any sources on the
+        # edge.
+        targetEpochs[~np.isfinite(targetEpochs)] = sourceMjd
+
+        with Matcher(isolatedSources["coord_ra"], isolatedSources["coord_dec"]) as m:
+            idx, i1, i2, d = m.query_radius(
+                properMotionCatalog["starX"], properMotionCatalog["starY"], 0.2 / 3600, return_indices=True
+            )
```

Comment:
This looks very similar to the code in `applyProperMotionCorrections`: can you consolidate some code between that and this?

Reply:
This is `applyProperMotionCorrections`...can you clarify?

Reply:
Ah, sorry, I meant `prepareAssociatedSources`: I was confused about which one I was looking at, because they were similar.

Reply:
I looked at this a bit, but I don't think there's much that could be consolidated.